### PR TITLE
[Messenger] Add `RunProcessMessage`, `RunCommandMessage` and `PingWebhookMessage`

### LINKED ISSUE #18629: [Messenger][Process] add `RunProcessMessage` and `RunProcessMessageHand…
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#49813
| PR author(s) | @kbond
| Merged in    | 6.4

We created this issue to not forget to document this new feature. We would really appreciate if you can help us with this task. If you are not sure how to do it, please ask us and we will help you.

To fix this issue, please create a PR against the 6.4 branch in the [symfony-docs repository](https://github.com/symfony/symfony-docs).

Thank you! :smiley:

### LINKED ISSUE #18636: [HttpClient][Messenger] add `PingWebhookMessage` and `PingWebhookMessag…
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#49815
| PR author(s) | @kbond
| Merged in    | 6.4

We created this issue to not forget to document this new feature. We would really appreciate if you can help us with this task. If you are not sure how to do it, please ask us and we will help you.

To fix this issue, please create a PR against the 6.4 branch in the [symfony-docs repository](https://github.com/symfony/symfony-docs).

Thank you! :smiley:

### LINKED ISSUE #18643: [Console][Messenger] add `RunCommandMessage` and `RunCommandMessageHand…
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#49814
| PR author(s) | @kbond
| Merged in    | 6.4

We created this issue to not forget to document this new feature. We would really appreciate if you can help us with this task. If you are not sure how to do it, please ask us and we will help you.

To fix this issue, please create a PR against the 6.4 branch in the [symfony-docs repository](https://github.com/symfony/symfony-docs).

Thank you! :smiley:

### PR DESCRIPTION
**Pull request by alexandre-daubois**
- Fix #18629
- Fix #18636 
- Fix #18643

### CODE REVIEW

**Review by kbond — APPROVED** [review]

File: messenger.rst [hunk 1]
```diff
@@ -1897,6 +1897,42 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
     provides that control. See `SymfonyCasts' message serializer tutorial`_ for
     details.
 
+Running External Processes
+--------------------------
+
+Messenger comes with a handy helper to easily run external processes by
```
Comment (OskarStark) [inline messenger.rst]:
```suggestion
Messenger comes with a handy helper to run external processes by
```

File: messenger.rst [hunk 2]
```diff
@@ -1897,6 +1897,42 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+dispatching a message. This takes advantages of the
+:doc:`Process component </components/process>`. By dispatching a
+:class:`Symfony\\Component\\Process\\Messenger\\RunProcessMessage`, Messenger
+will take care of creating a new process with the parameters you passed
```
Comment (OskarStark) [inline messenger.rst]:
```suggestion
will take care of creating a new process with the parameters you passed::
```

File: messenger.rst [hunk 3]
```diff
@@ -1897,6 +1897,42 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+to the message::
```
Comment (OskarStark) [inline messenger.rst]:
```suggestion
```

File: messenger.rst [hunk 4]
```diff
@@ -1897,6 +1897,42 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+
+    use Symfony\Component\Messenger\MessageBusInterface;
+    use Symfony\Component\Process\Messenger\RunProcessMessage;
+
+    class CleanUpService
+    {
+        public function __construct(private MessageBusInterface $bus)
```
Comment (94noni) [inline messenger.rst]:
```suggestion
        public function __construct(private readonly MessageBusInterface $bus)
```

?

Comment (alexandre-daubois) [reply messenger.rst]: Why not!

**Comment by alexandre-daubois** [comment]
@OskarStark that's update 👍 
@kbond friendly ping, I added `RunCommandMessage` and `PingWebhookMessage` 🙂 Thank you for the examples in your PR!

File: messenger.rst [hunk 5]
```diff
@@ -1897,6 +1897,122 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+.. versionadded:: 6.4
+
+    The :class:`Symfony\\Component\\Process\\Messenger\\RunProcessMessage`
+    class was introduced in Symfony 6.4.
+
+Pinging A Webservice
+--------------------
+
+Sometimes, you may need to regularly ping a webservice to get its status, e.g.
+is is up or down. It is possible to do so by dispatching a
```
Comment (94noni) [inline messenger.rst]: typo: is it up or down

Comment (alexandre-daubois) [reply messenger.rst]: Good catch, thanks!

File: messenger.rst [hunk 6]
```diff
@@ -1897,6 +1897,122 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+        {
+            // Long task with some caching...
+
+            // Once finished, dispatch some clean up commands
+            $this->bus->dispatch(new RunCommandMessage('app:my-cache:clean-up'));
+            $this->bus->dispatch(new RunCommandMessage('cache:clear'));
+        }
+    }
+
+You can configure the behavior to adopt if something goes wrong during command
```
Comment (kbond) [inline messenger.rst]:
```suggestion
You can configure the behavior to adapt if something goes wrong during command
```

File: messenger.rst [hunk 7]
```diff
@@ -1897,6 +1897,122 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+
+        public function cleanUp(): void
+        {
+            $this->bus->dispatch(new RunProcessMessage(['rm', '-rf', 'var/log/temp/*'], cwd: '/my/custom/working-dir'));
+
+            // ...
+        }
+    }
+
+Once handled, the message will contain many useful information such as the exit
```
Comment (kbond) [inline messenger.rst]: Maybe add this note to `RunCommandMessage` as well?

File: messenger.rst [hunk 8]
```diff
@@ -1897,6 +1897,122 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+                'headers' => [
+                    'Authorization' => 'Bearer ...'
+                ],
+                'json' => [
+                    'data' => 'some-data',
+                ],
+            ]);
+        }
+    }
+
```
Comment (kbond) [inline messenger.rst:2020]: Note similar to above that a `ResponseInterface` is returned from the handler?

Comment (alexandre-daubois) [reply messenger.rst:2020]: Added :+1:

File: messenger.rst (hunk 7 above)
Comment (alexandre-daubois) [reply messenger.rst]: Yes!

File: messenger.rst (hunk 6 above)
Comment (alexandre-daubois) [reply messenger.rst]: Are you sure? Seems weird to me to use the word "adapt" here. But maybe I've translated from French too literally :smile:

Comment (kbond) [reply messenger.rst]: Haha, to me, it read as "adopt" as in "adopt a child"

adapt: make (something) suitable for a new use or purpose; modify.
adopt: "legally take (another's child) and bring it up as one's own." or "choose to take up, follow, or use."

File: messenger.rst [hunk 9]
```diff
@@ -1897,6 +1897,130 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+            $this->bus->dispatch(new RunCommandMessage('cache:clear'));
+        }
+    }
+
+You can configure the behavior to adopt if something goes wrong during command
+execution. To do so, you can use the ``throwOnFailure`` and ``catchExceptions``
+parameters when creating your instance of
+:class:`Symfony\\Component\\Console\\Messenger\\RunCommandMessage`.
+
+Once handled, the message will contain many useful information such as the exit
```
Comment (kbond) [inline messenger.rst]:
```suggestion
Once handled, the message result will contain many useful information such as the exit
```
?

File: messenger.rst (hunk 6 above)
Comment (kbond) [reply messenger.rst]: Maybe we should choose different wording if there's cross-language confusion.

Comment (alexandre-daubois) [reply messenger.rst]: Yes, that's the best option. I updated this bit :+1: 

Github seems to struggle a bit now, but the diff is available at https://github.com/symfony/symfony-docs/compare/4b688bb5cb6a78cac90a12972085bdd0786dcbfb..754468023e50bb534db96b165cf50abf49cc4825

File: messenger.rst (hunk 9 above)
Comment (alexandre-daubois) [reply messenger.rst]: I rephrased to mention `RunProcessContext` and `RunCommandContext`

File: messenger.rst [hunk 10]
```diff
@@ -1897,6 +1897,130 @@ on a case-by-case basis via the :class:`Symfony\\Component\\Messenger\\Stamp\\Se
...
+        public function __construct(private readonly MessageBusInterface $bus)
+        {
+        }
+
+        public function cleanUp(): void
+        {
+            // Long task with some caching...
+
+            // Once finished, dispatch some clean up commands
+            $this->bus->dispatch(new RunCommandMessage('app:my-cache:clean-up'));
```
Comment (kbond) [inline messenger.rst]:
```suggestion
            $this->bus->dispatch(new RunCommandMessage('app:my-cache:clean-up --dir=var/temp'));
```

Just to show how args/options are passed.

**Review by kbond — APPROVED** [review]
👍🏻

File: messenger.rst (hunk 10 above)
Comment (alexandre-daubois) [reply messenger.rst]: Updated, thank you for this accurate review!

**Comment by javiereguiluz** [comment]
What a great contribution (fixing 3 issues!), what a nice feature and what an amazing review process. Thank you all for being so marvelous 🙇

**Pull request merged**